insights/parsers/smartctl.py

```diff
@@ -1,55 +1,111 @@
+          4 Start_Stop_Count        0x0032   100   100   020    Old_age   Always       -       546
+          5 Reallocated_Sector_Ct   0x0033   100   100   036    Pre-fail  Always       -       0
+        ...
+
+    Examples:
+        >>> for drive in shared[SMARTctl]:
+        ...     print "Device:", drive.device
+        ...     print "Model:", drive.info['Device Model']
+        ...     print "Health check:", drive.health
+        ...     print "Last self-test status:", drive.values['Self-test execution status']
+        ...     print "Raw read error rate:", drive.attributes['Raw_Read_Error_Rate']['RAW_VALUE']
+        ...
+        Device: /dev/sda
+        Model: ST500LM021-1KJ152
+        Health check: PASSED
+        Last self-test status: 0
+        Raw read error rate: 179599704
+
     """
 
-    INFO_LINE_STR = r'(?P<key>\w+(?:\s\w+)*):\s+' + \
+    _INFO_LINE_STR = r'(?P<key>\w+(?:\s\w+)*):\s+' + \
         r'(?P<value>\S.*?)\s*$'
-    INFO_LINE_RE = re.compile(INFO_LINE_STR)
-    VALUE_LINE_STR = r'(?P<key>\w[A-Za-z _.-]+):\s+' + \
+    _INFO_LINE_RE = re.compile(_INFO_LINE_STR)
+    _VALUE_LINE_STR = r'(?P<key>\w[A-Za-z _.-]+):\s+' + \
         r'\(\s*(?P<value>\S.*?)\)'
-    VALUE_LINE_RE = re.compile(VALUE_LINE_STR)
-    ATTR_LINE_STR = r'^\s*(?P<id>\d+)\s(?P<name>\w+)\s+' + \
+    _VALUE_LINE_RE = re.compile(_VALUE_LINE_STR)
+    _ATTR_LINE_STR = r'^\s*(?P<id>\d+)\s(?P<name>\w+)\s+' + \
         r'(?P<flag>0x[0-9a-fA-F]{4})\s+(?P<value>\d{3})\s+' + \
         r'(?P<worst>\d{3})\s+(?P<threshold>\d{3})\s+' + \
         r'(?P<type>[A-Za-z_-]+)\s+(?P<updated>[A-Za-z_-]+)\s+' + \
         r'(?P<when_failed>\S+)\s+(?P<raw_value>\S.*)$'
-    ATTR_LINE_RE = re.compile(ATTR_LINE_STR)
-    ATTR_KEYS = ['id', 'flag', 'value', 'worst', 'threshold', 'type',
-                 'updated', 'when_failed', 'raw_value']
+    _ATTR_LINE_RE = re.compile(_ATTR_LINE_STR)
 
     def __init__(self, context):
         super(SMARTctl, self).__init__(context)
         filename_re = re.compile(r'smartctl_-a_\.dev\.(?P<device>\w+)$')
         match = filename_re.search(context.path)
         self.data = {}
         if match:
-            self.device = match.group('device')
-        # Else warn?  fail?
+            self.device = '/dev/' + match.group('device')
+        else:
+            raise ParseException('Cannot parse device name from path {p}'.format(p=context.path))
 
     def parse_content(self, content):
-        drive_info = {
-            'info': {},
-            'health': 'not parsed',
-            'values': {},
-            'attributes': {},
-            'full_line': '',  # hack for persistent line storage
-        }
+        self.info = {}
+        self.health = 'not parsed'
+        self.values = {}
+        self.attributes = {}
+        # hack for persistent line storage in parse_content context -
+        # otherwise it gets treated as a local variable within the sub-
+        # functions
+        self.full_line = ''
```

Comment:
@PaulWay while we may not be using this on our projects, we don't know how it is being used by other internal or external users.  So we need to minimize the impact to those users we may not know about now that the project is publicly available.  You can make all of these changes on the master branch but we need to maintain backward compatibility on this branch.